### PR TITLE
UI: VAULT-17008 Dashboard Secrets Engines Card

### PR DESCRIPTION
**Pull request by kiannaquach**
**Description**:
- Query data for the secrets engines card
- Create component for secrets engines card
- Create secrets engines card test
- TODO: Fix the table to have rounded borders (VAULT-17365)
- TODO for design
    - ask yongqi about the Secrets Engines Title so that it's more specific to users about how many are being shown
    - ask about namespaces and whether we should move them up or include them in dashboard header
    - ask about how unsupported backends look in the card
    - confirm how long descriptions will look
Screenshot 2023-06-20 at 12 30 29 PM

### CODE REVIEW

File: ui/app/components/dashboard/secrets-engines-card.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import Component from '@glimmer/component';
+
+export default class DashboardSecretsEnginesCardComponent extends Component {
```
Comment (Monkeychip) [inline ui/app/components/dashboard/secrets-engines-card.js:8]: NIT: you don't need to end the name of the component with "Component". Similar to `KeymgmtDistribute` component, it's named the file path without the end of the component. Curious, did this naming come default from the blueprints? So name would be `DashboardSecretsEnginesCard`

File: ui/app/components/dashboard/secrets-engines-card.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import Component from '@glimmer/component';
+
```
Comment (Monkeychip) [inline ui/app/components/dashboard/secrets-engines-card.js:7]: I realize we are so very not consistent with this, but could you add documentation to the js file here? You can see documentation for example in `app/components/keymgmt/distribute`. Documentation is most helpful when you're going to reuse a component and pass in params, so technically this maybe optional here, but I also find it helpful when understanding what the file is going to show.

File: ui/app/components/dashboard/secrets-engines-card.js (hunk 1 above)
Comment (Monkeychip) [reply ui/app/components/dashboard/secrets-engines-card.js:8]: Though I do see at least one use case where we do end in component. PolicyFormComponent. So ignore, I think it's a personal preference. Doesn't look like a hard fast rule.

File: ui/app/routes/vault/cluster/secrets/backends.js [hunk 3]
```diff
@@ -10,6 +10,8 @@ export default Route.extend({
   store: service(),
 
   model() {
+    // TODO LANDING PAGE: VAULT-17008 use peekAll to avoid a network request
```
Comment (Monkeychip) [inline ui/app/routes/vault/cluster/secrets/backends.js:13]: :)

File: ui/app/styles/helper-classes/flexbox-and-grid.scss [hunk 4]
```diff
@@ -18,6 +18,10 @@
 .is-flex-row {
   display: flex;
   flex-direction: row;
+
+  &.has-gap {
```
Comment (Monkeychip) [inline ui/app/styles/helper-classes/flexbox-and-grid.scss:22]: :)

File: ui/app/styles/helper-classes/flexbox-and-grid.scss [hunk 5]
```diff
@@ -18,6 +18,10 @@
 .is-flex-row {
   display: flex;
   flex-direction: row;
+
```
Comment (Monkeychip) [inline ui/app/styles/helper-classes/flexbox-and-grid.scss:21]: My goal with helpers is they do one thing. Would you mind pulling this out into it's own helper called `gap-m` (we're working on dropping the starting verb).You can put it anywhere in this helper file you think would be good. 

And I know, I have a `has-gap` on row 59 of this file. It was a carry over and though I hated it, I did the least amount of overhaul I could during the bulma removal, which meant copying over old patterns. 

In fact, if you add that helper to this file, we can then also remove the has-gap usage and replace with that helper.

File: ui/app/templates/components/dashboard/secrets-engines-card.hbs [hunk 6]
```diff
@@ -0,0 +1,69 @@
+<div class="is-flex-between has-left-margin-xs" data-test-dashboard-secrets-engines-header>
+  <h3 class="title is-4">Secrets Engines</h3>
+  <div>
+    <Hds::Link::Inline @route="vault.cluster.secrets.backends" class="has-text-weight-semibold is-no-underline">
+      View more
+    </Hds::Link::Inline>
+  </div>
+</div>
+<Hds::Table @caption="Five secrets engines" class="has-border-collapse" data-test-dashboard-secrets-engines-table>
```
Comment (Monkeychip) [inline ui/app/templates/components/dashboard/secrets-engines-card.hbs:9]: As mentioned, not everyone is going to have five secrets engines. They may just have one. Not sure what to replace this caption with, but something to talk with design about?

Also, maybe I'm missing something, but is this caption supposed to show?
image

File: ui/app/templates/components/dashboard/secrets-engines-card.hbs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                  <span class="has-text-grey" data-test-secret-path>{{backend.path}}</span>
+                {{/if}}
+              {{/if}}
+            </div>
+            {{#if backend.accessor}}
+              <code class="has-text-grey is-size-8" data-test-accessor>
+                {{backend.accessor}}
+              </code>
+            {{/if}}
+            {{#if backend.description}}
```
Comment (Monkeychip) [inline ui/app/templates/components/dashboard/secrets-engines-card.hbs:48]: I'm not sure how we handle it on the normal secret engine list view, but try this out with a long description. It get's cutoff. You can add description under options when you first mount a secret engine. 
image

image

Comment (Monkeychip) [reply ui/app/templates/components/dashboard/secrets-engines-card.hbs:48]: Ahh, I think Claire worked on this component. But we add a "See more" on the regular view. Probably too many links for this card, but maybe a `...` after a certain string length. Design might have some ideas. 
image

File: ui/tests/acceptance/dashboard-test.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    await visit('/vault/dashboard');
+
+    assert.strictEqual(currentURL(), '/vault/dashboard');
+  });
+
+  test('shows the secrets engines card', async function (assert) {
+    await enablePage.enable('pki', 'pki');
+    await visit('/vault/dashboard');
+    assert.dom(SELECTORS.cardTitle).hasText('Secrets Engines');
+    assert.dom(SELECTORS.getSecretEngineAccessor('pki')).exists();
```
Comment (Monkeychip) [inline ui/tests/acceptance/dashboard-test.js:33]: when you get the disabled class styling setup, maybe check that the item has unsupported styling? You can use the `engines-test` as an example. Specifically, `it adds disabled css styling to unsupported secret engines`

**Review by Monkeychip — APPROVED** [review]
A couple of comments and things we chatted about via zoom. Given this is a sidebranch and we're waiting to clarify some things with design (whose OOO for right now), I'll go ahead and approve so that you're not blocked.

File: ui/app/styles/helper-classes/flexbox-and-grid.scss (hunk 5 above)
Comment (kiannaquach) [reply ui/app/styles/helper-classes/flexbox-and-grid.scss:21]: Good point - will update that!

File: ui/app/components/dashboard/secrets-engines-card.js (hunk 1 above)
Comment (kiannaquach) [reply ui/app/components/dashboard/secrets-engines-card.js:8]: I do see the redundancy - I will remove it! I think we added Component to the end in the PKI and the blueprint automatically creates the name with the Component in the end.

File: ui/app/templates/components/dashboard/secrets-engines-card.hbs (hunk 6 above)
Comment (kiannaquach) [reply ui/app/templates/components/dashboard/secrets-engines-card.hbs:9]: @Monkeychip Which caption are you referencing? The Vault version will be tackled on a separate ticket, but the description for the secret engine is supposed to show :)

File: ui/app/templates/components/dashboard/secrets-engines-card.hbs (hunk 7 above)
Comment (kiannaquach) [reply ui/app/templates/components/dashboard/secrets-engines-card.hbs:48]: Good catch! I will ask Yongqi about this. For now, I will display the whole description and will ask Yongqi about how we want to tackle long descriptions.


Screenshot 2023-06-20 at 1 46 38 PM

File: ui/app/templates/components/dashboard/secrets-engines-card.hbs (hunk 6 above)
Comment (kiannaquach) [reply ui/app/templates/components/dashboard/secrets-engines-card.hbs:9]: Just left a comment on Figma for design!

**Pull request merged**